query/src/servers/http/v1/clickhouse_handler.rs

```diff
@@ -0,0 +1,225 @@
+
+    let data_stream = interpreter.execute(input_stream).await?;
+    let mut data_stream = ctx.try_create_abortable(data_stream)?;
+
+    let stream = stream! {
+           while let Some(block) = data_stream.next().await {
+            match block{
+                Ok(block) => {
+                    yield(block_to_tsv(&block))
+                },
+                Err(err) => yield(Err(err)),
+            };
+        }
+    };
+
+    Ok(Body::from_bytes_stream(stream))
+}
+
+#[poem::handler]
+pub async fn clickhouse_handler_get(
+    sessions_extension: Data<&Arc<SessionManager>>,
+    user_info: Data<&UserInfo>,
+    Query(params): Query<StatementHandlerParams>,
+) -> PoemResult<Body> {
+    let session_manager = sessions_extension.0;
+    let session = session_manager
+        .create_session(SessionType::ClickHouseHttpHandler)
+        .await
+        .map_err(InternalServerError)?;
+    session.set_current_user(user_info.0.clone());
+
+    let context = session
+        .create_query_context()
+        .await
+        .map_err(InternalServerError)?;
+
+    let mut sql = params.query;
+    let plan = PlanParser::parse(context.clone(), &sql)
+        .await
+        .map_err(BadRequest)?;
+
+    if matches!(plan, PlanNode::Insert(_)) {
+        return Err(BadRequest(ErrorCode::SyntaxException(
+            "not allow insert in GET",
+        )));
+    }
+    context.attach_query_str(&sql);
+    execute(context, plan, None)
+        .await
+        .map_err(InternalServerError)
+}
+
+fn try_parse_insert_formatted(sql: &str) -> Result<Option<(Format, Vec<DfStatement>)>> {
+    if let Ok((statements, _)) = DfParser::parse_sql(sql) {
+        if statements.is_empty() {
+            return Ok(None);
+        }
+        if statements.len() > 1 {
+            return Err(ErrorCode::SyntaxException("Only support single query"));
+        }
+        let statement = &statements[0];
+        if let DfStatement::InsertQuery(ref insert) = statement {
+            if let Some(format) = &insert.format {
+                match format.as_str() {
+                    FORMAT_JSON_EACH_ROW => return Ok(Some((Format::NDJson, statements))),
+                    // "" for "insert into my_table values;"
+                    "" => {}
+                    _ => {
+                        return Err(ErrorCode::SyntaxException(format!(
+                            "format {} not supported; only support: {}",
+                            format,
+                            supported_formats()
+                        )))
+                    }
+                };
+            }
+        };
+    }
+    Ok(None)
+}
+
+// parse query params first to avoid tokenize the body part when it is a insert with format.
+async fn parse(
+    sql: &str,
+    body: Body,
+) -> PoemResult<(Option<(Format, Body)>, Cow<'_, str>, Vec<DfStatement>)> {
+    if let Some((format, statements)) = try_parse_insert_formatted(sql).map_err(BadRequest)? {
```

Comment:
If it is not insert SQL, it will be parsed twice.
I think it's better to reuse the `statements`.

Reply:
@sundy-li statements are already reused now, "parsed twice" only happens when need to append the body part.

Reply:
for POST, parse the 'query' parameter first in case it is "insert ...  format".
if not (parse fail or not insert format ),  append the body and parse again.

Reply:
I thought about reusing the tokens,  but since the  'query' parameter is not long most of the time, I wonder if it is worth it.

Reply:
It's trivial.

Yet when the SQL is very long, users may post a select query inside the body, then it's parsed twice.

I suggest to call :

```
let (statements, _) = DfParser::parse_sql(&sql).map_err(BadRequest)?;
```

And match the statements to get the format.

Reply:
> sql



> It's trivial.
> 
> Yet when the SQL is very long, users may post a select query inside the body, then it's parsed twice.
> 
> I suggest to call :
> 
> ```
> let (statements, _) = DfParser::parse_sql(&sql).map_err(BadRequest)?;
> ```
> 
> And match the statements to get the format.

what is the `sql`?   including the POST body part?

Reply:
ClickHouse supported  post the sql into body instead of query params.

```
➜ curl http://127.0.0.1:8123 -d 'select * from numbers(10)'
0
1
2
3
4
5
6
7
8
9
```

Reply:
we already support it now: https://github.com/datafuselabs/databend/pull/4613/files#diff-cf80e97711c9701ffe83aa5deded1afa6581db50518d4b25548fec8bb4d84507R79

Reply:
the problem is that the current DfParser tokenize the whole query at first (get a vec<Token> instead of gen a token stream)
so it will be a bigger waste to parse`"query" + "body" ` if the body is a big data file.